.github/workflows/crucible-ci.yaml

```diff
@@ -9,29 +9,83 @@ on:
   workflow_dispatch:
 
 jobs:
-  crucible-ci:
-    runs-on: [self-hosted, cpu-partitioning, remotehost]
+  generate-job-matrix-parameters:
+    runs-on: ubuntu-latest
+    outputs:
+      userenvs: ${{ steps.get-userenvs.outputs.userenvs }}
+      endpoints_self: ${{ steps.get-endpoints-self.outputs.endpoints }}
+    steps:
+    - name: checkout crucible-ci
+      uses: actions/checkout@v2
+      with:
+        repository: perftool-incubator/crucible-ci
+        ref: main
+        path: crucible-ci
+    - name: checkout rickshaw
+      uses: actions/checkout@v2
+      with:
+        repository: perftool-incubator/rickshaw
+        ref: master
+        path: rickshaw
+    - name: run get-userenvs
+      id: get-userenvs
+      uses: ./crucible-ci/.github/actions/get-userenvs
+      with:
+        rickshaw-directory: "./rickshaw"
+    - name: run get-endpoints-self
+      id: get-endpoints-self
+      uses: ./crucible-ci/.github/actions/get-endpoints
+      with:
+        runner-type: "self"
+
+  display-job-matrix-parameters:
+    runs-on: ubuntu-latest
+    needs: generate-job-matrix-parameters
+    steps:
+    - name: Echo generate-job-matrix-parameters outputs
+      run: echo "${{ toJSON(needs.generate-job-matrix-parameters.outputs) }}"
 
+  crucible-ci-self-hosted-runners:
+    runs-on: [ self-hosted ]
+    needs:
+    - generate-job-matrix-parameters
+    - display-job-matrix-parameters
+    continue-on-error: true
+    strategy:
+      matrix:
+        endpoint: ${{ fromJSON(needs.generate-job-matrix-parameters.outputs.endpoints_self) }}
+        userenv: ${{ fromJSON(needs.generate-job-matrix-parameters.outputs.userenvs) }}
```

Comment:
does this mean these steps will be executed in parallel for all endpoints ?

Reply:
It is kinda like multiplex, it takes the arrays defined under a strategy (named 'matrix') and creates a job for all possible combinations.  The jobs are all queued up at the same time (somewhat, it takes a bit longer to create them than you might think, on the order of 10s of seconds).  How many actually run simultaneously is governed by the Github runner limits, for our `perftool-incubator` project that is 20.